chia/consensus/mmr.py

```diff
@@ -0,0 +1,279 @@
+from __future__ import annotations
+
+import logging
+
+from chia_rs.sized_bytes import bytes32
+from chia_rs.sized_ints import uint32
+
+from chia.util.hash import std_hash
+from chia.util.streamable import Streamable
+
+log = logging.getLogger(__name__)
+
+
+# ------------------------------------------------------------------------------
+# MMR position/height utilities
+# ------------------------------------------------------------------------------
+
+
+def get_height(flat_index: int) -> int:
+    """
+    Calculate the height of a node in a flat MMR array.
+    """
+    x = flat_index + 1  # Work with 1-based for easier math
+
+    while True:
+        # Check if x is "all ones" (1, 3, 7, 15...) -> Peak of perfect binary tree
+        if (x & (x + 1)) == 0:
+            return x.bit_length() - 1
+
+        # Not a peak, subtract left sibling mountain size
+        msb_val = 1 << (x.bit_length() - 1)
+        subtree_size = msb_val - 1
+        x -= subtree_size
+
+
+def get_peak_positions(size: int) -> list[int]:
+    """
+    Identify the indices of the mountain peaks.
+    Returns indices [Rightmost Peak (Smallest), ... Leftmost Peak (Tallest)]
+    """
+    peaks = []
+    idx = size - 1
+
+    while idx >= 0:
+        peaks.append(idx)
+        height = get_height(idx)
+        # Size of this mountain = 2^(h+1) - 1
+        mountain_size = (1 << (height + 1)) - 1
+        idx -= mountain_size
+
+    return peaks
+
+
+def leaf_index_to_pos(leaf_index: int) -> int:
+    """
+    Convert a leaf index (0-based) to its position in the flat MMR.
+    Formula: 2*L - popcount(L)
+    """
+    return 2 * leaf_index - bin(leaf_index).count("1")
+
+
+# ------------------------------------------------------------------------------
+# Class Implementation
+# ------------------------------------------------------------------------------
+
+
+class MerkleMountainRange(Streamable):
+    """
+    Flat MMR implementation.
+    """
+
+    nodes: list[bytes32]
```

Comment:
I think it would be an important performance improvement to make this a `bytearray` instead. But these kinds of optimizations can wait.